File: backend/src/AgentTeam/Models/Team.php
<?php

declare(strict_types=1);

namespace AgentTeam\Models;

/**
 * Team Model
 *
 * Represents a team that groups agents together.
 *
 * @see docs/agentDesign.md
 */
class Team
{
    private ?int $id = null;
    private int $userId = 0;
    private ?int $workspaceId = null;
    private string $name = '';
    private string $description = '';
    private ?string $createdAt = null;
    private ?string $updatedAt = null;

    // Related agents (loaded separately)
    private array $agents = [];

    // Agents with pipeline info (position, labels, etc.)
    private array $agentsWithPipelineInfo = [];

    /**
     * Create a new Team instance
     */
    public function __construct(array $data = [])
    {
        if (!empty($data)) {
            $this->hydrate($data);
        }
    }

    /**
     * Hydrate the team from an array (e.g., database row)
     */
    public function hydrate(array $data): self
    {
        $this->id = isset($data['id']) ? (int) $data['id'] : null;
        $this->userId = isset($data['user_id']) ? (int) $data['user_id'] : 0;
        $this->workspaceId = isset($data['workspace_id']) ? (int) $data['workspace_id'] : null;
        $this->name = $data['name'] ?? '';
        $this->description = $data['description'] ?? '';
        $this->createdAt = $data['created_at'] ?? null;
        $this->updatedAt = $data['updated_at'] ?? null;

        return $this;
    }

    /**
     * Convert the team to an array
     */
    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'user_id' => $this->userId,
            'workspace_id' => $this->workspaceId,
            'name' => $this->name,
            'description' => $this->description,
            'created_at' => $this->createdAt,
            'updated_at' => $this->updatedAt,
        ];
    }

    /**
     * Convert to array with agents included (with pipeline info)
     */
    public function toArrayWithAgents(): array
    {
        $data = $this->toArray();

        // If we have agents with pipeline info, use that
        if (!empty($this->agentsWithPipelineInfo)) {
            $data['agents'] = array_map(function ($item) {
                $agentData = $item['agent']->toApiArray();
                $agentData['pipeline_position'] = $item['pipeline_position'];
                $agentData['pipeline_label'] = $item['pipeline_label'];
                $agentData['is_pipeline_head'] = $item['is_pipeline_head'];
                $agentData['is_pipeline_tail'] = $item['is_pipeline_tail'];
                $agentData['can_reorder'] = $item['can_reorder'];
                return $agentData;
            }, $this->agentsWithPipelineInfo);
        } else {
            // Fallback to basic agents array
            $data['agents'] = array_map(
                fn($agent) => $agent->toApiArray(),
                $this->agents
            );
        }

        return $data;
    }

    /**
     * Convert to array for JSON API response
     */
    public function toApiArray(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'description' => $this->description,
            'created_at' => $this->createdAt,
        ];
    }

    // ========================================
    // Getters
    // ========================================

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getUserId(): int
    {
        return $this->userId;
    }

    public function getWorkspaceId(): ?int
    {
        return $this->workspaceId;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getDescription(): string
    {
        return $this->description;
    }

    public function getCreatedAt(): ?string
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): ?string
    {
        return $this->updatedAt;
    }

    public function getAgents(): array
    {
        return $this->agents;
    }

    // ========================================
    // Setters (Fluent Interface)
    // ========================================

    public function setId(int $id): self
    {
        $this->id = $id;
        return $this;
    }

    public function setUserId(int $userId): self
    {
        $this->userId = $userId;
        return $this;
    }

    public function setWorkspaceId(?int $workspaceId): self
    {
        $this->workspaceId = $workspaceId;
        return $this;
    }

    public function setName(string $name): self
    {
        $this->name = $name;
        return $this;
    }

    public function setDescription(string $description): self
    {
        $this->description = $description;
        return $this;
    }

    public function setAgents(array $agents): self
    {
        $this->agents = $agents;
        return $this;
    }

    /**
     * Set agents with pipeline info (position, labels, etc.)
     *
     * @param array $agentsWithInfo Array of ['agent' => Agent, 'pipeline_position' => int|null, ...]
     */
    public function setAgentsWithPipelineInfo(array $agentsWithInfo): self
    {
        $this->agentsWithPipelineInfo = $agentsWithInfo;

        // Also populate basic agents array for backwards compatibility
        $this->agents = array_map(fn($item) => $item['agent'], $agentsWithInfo);

        return $this;
    }

    public function getAgentsWithPipelineInfo(): array
    {
        return $this->agentsWithPipelineInfo;
    }
}

File: backend_python/app/agent_team/models/team.py
"""Port of backend/src/AgentTeam/Models/Team.php.

Represents a team that groups agents together.
"""
from __future__ import annotations

from app.support.phpcompat import php_intval


class Team:
    def __init__(self, data: dict | None = None):
        """PHP: `if (!empty($data)) { $this->hydrate($data); }`."""
        self.id: int | None = None
        self.userId: int = 0
        self.workspaceId: int | None = None
        self.name: str = ''
        self.description: str = ''
        self.createdAt: str | None = None
        self.updatedAt: str | None = None

        # Related agents (loaded separately)
        self.agents: list = []

        # Agents with pipeline info (position, labels, etc.)
        self.agentsWithPipelineInfo: list = []

        if data:
            self.hydrate(data)

    def hydrate(self, data: dict) -> 'Team':
        self.id = php_intval(data['id']) if data.get('id') is not None else None
        self.userId = php_intval(data['user_id']) if data.get('user_id') is not None else 0
        self.workspaceId = php_intval(data['workspace_id']) if data.get('workspace_id') is not None else None
        self.name = data.get('name') if data.get('name') is not None else ''
        self.description = data.get('description') if data.get('description') is not None else ''
        self.createdAt = data.get('created_at') if data.get('created_at') is not None else None
        self.updatedAt = data.get('updated_at') if data.get('updated_at') is not None else None

        return self

    def toArray(self) -> dict:
        return {
            'id': self.id,
            'user_id': self.userId,
            'workspace_id': self.workspaceId,
            'name': self.name,
            'description': self.description,
            'created_at': self.createdAt,
            'updated_at': self.updatedAt,
        }

    def toArrayWithAgents(self) -> dict:
        data = self.toArray()

        # If we have agents with pipeline info, use that
        if self.agentsWithPipelineInfo:
            agents = []
            for item in self.agentsWithPipelineInfo:
                agent_data = item['agent'].toApiArray()
                agent_data['pipeline_position'] = item['pipeline_position']
                agent_data['pipeline_label'] = item['pipeline_label']
                agent_data['is_pipeline_head'] = item['is_pipeline_head']
                agent_data['is_pipeline_tail'] = item['is_pipeline_tail']
                agent_data['can_reorder'] = item['can_reorder']
                agents.append(agent_data)
            data['agents'] = agents
        else:
            # Fallback to basic agents array
            data['agents'] = [agent.toApiArray() for agent in self.agents]

        return data

    def toApiArray(self) -> dict:
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'created_at': self.createdAt,
        }

    # ========================================
    # Getters
    # ========================================

    def getId(self) -> int | None:
        return self.id

    def getUserId(self) -> int:
        return self.userId

    def getWorkspaceId(self) -> int | None:
        return self.workspaceId

    def getName(self) -> str:
        return self.name

    def getDescription(self) -> str:
        return self.description

    def getCreatedAt(self) -> str | None:
        return self.createdAt

    def getUpdatedAt(self) -> str | None:
        return self.updatedAt

    def getAgents(self) -> list:
        return self.agents

    # ========================================
    # Setters (Fluent Interface)
    # ========================================

    def setId(self, id: int) -> 'Team':
        self.id = id
        return self

    def setUserId(self, user_id: int) -> 'Team':
        self.userId = user_id
        return self

    def setWorkspaceId(self, workspace_id: int | None) -> 'Team':
        self.workspaceId = workspace_id
        return self

    def setName(self, name: str) -> 'Team':
        self.name = name
        return self

    def setDescription(self, description: str) -> 'Team':
        self.description = description
        return self

    def setAgents(self, agents: list) -> 'Team':
        self.agents = agents
        return self

    def setAgentsWithPipelineInfo(self, agents_with_info: list) -> 'Team':
        """agentsWithInfo: list of {'agent': Agent, 'pipeline_position': int|None, ...}."""
        self.agentsWithPipelineInfo = agents_with_info

        # Also populate basic agents array for backwards compatibility
        self.agents = [item['agent'] for item in agents_with_info]

        return self

    def getAgentsWithPipelineInfo(self) -> list:
        return self.agentsWithPipelineInfo
